python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -529,15 +524,15 @@ def recover_if_needed(self, node_id, now):
         if not self.can_update(node_id):
             return
         key = self.provider.internal_ip(node_id)
-        if key not in self.load_metrics.last_heartbeat_time_by_ip:
-            self.load_metrics.last_heartbeat_time_by_ip[key] = now
-        last_heartbeat_time = self.load_metrics.last_heartbeat_time_by_ip[key]
-        delta = now - last_heartbeat_time
-        if delta < AUTOSCALER_HEARTBEAT_TIMEOUT_S:
-            return
+        if key in self.load_metrics.last_heartbeat_time_by_ip:
```

Comment:
This was a _very_ subtle latent bug. It meant that keys in `self.load_metrics.last_heartbeat_time_by_ip` were inconsistent with the other mappings.

Reply:
But you need another way to address this.
If you remove this, we will never remove nodes that had a crash in raylet.
What this does is that it stores the heartbeat immediately when raylet crashes on the node, so that after idle timeout it is removed.

Reply:
You might want to also add a name for these nodes.

Reply:
> If you remove this, we will never remove nodes that had a crash in raylet.

I don't think that's the existing behavior. This method should've _restarted_ crashed raylets, not terminate the node. I'm pretty sure it was just buggy because it was copied and pasted from the startup logic (where it _does_ make sense to prepopulate the mapping). 

> You might want to also add a name for these nodes.

What do you mean by a "name"? I don't think I've introduced anything here, just fixed a bug.

Reply:
I understand that it should have restarted crashed raylets, but what I am not sure about is what happens to heartbeatless nodes ? which cases originally required handling this case?

Reply:
If a raylet isn't sending heartbeats that's the same as a crashed raylet.

Reply:
OK, if raylet crashes, is it guaranteed to have an entry in `last_heartbeat_time_by_ip`?
If the answer is yes we can resolve this, if not, then how this restarts the crashed raylet?